### PR TITLE
Make `ExHal.Navigation.follow_links/4` concurrent.

### PR DESCRIPTION
**Pull request by lmarlow**
Following multiple links could make many requests to multiple servers,
so fetch them concurrently using `Task.async_stream/3`.

Anecdotally, this took a request to follow 40+ remote links from 8-9 seconds down to 2-3 seconds.

### CODE REVIEW

File: lib/exhal/navigation.ex [hunk 1]
```diff
@@ -42,7 +42,10 @@ defmodule ExHal.Navigation do
 
     case ExHal.get_links_lazy(a_doc, name, fn -> :missing end) do
       :missing -> missing_link_handler.(name)
-      links -> Enum.map(links, &_follow_link(a_doc.client, &1, tmpl_vars, opts))
+      links ->
+        links
+        |> Task.async_stream(&_follow_link(a_doc.client, &1, tmpl_vars, opts))
+        |> Enum.map(fn {:ok, followed_link} -> followed_link end)
```
Comment (pezra) [inline lib/exhal/navigation.ex]: `_follow_link/4` can fail. What happens if fetching one of the links is unable to connect, times out, returns a 5XX, returns 4XX, returns a non-HAL response?

Comment (pezra) [inline lib/exhal/navigation.ex]: Codebeat is complaining about the length of this function. What if we changed its signature to `follow_links(a_doc, name, missing_link_handler, tmpl_vars: tmpl_vars)` and moved the Map unpacking to another function head?

Comment (lmarlow) [reply lib/exhal/navigation.ex]: The underlying `@client_module.get/3` returns either `{:ok, _}` or `{:error, _}` which will just be unwrapped by the `Enum.map/2`. Should we handle the possibility of it crashing?

Comment (lmarlow) [reply lib/exhal/navigation.ex]: It seems like we still need to pass `opts` down to `@client_module` for headers and `Poison` options. I could simply extract the async and map into a private function to satisfy Codebeat.

Comment (pezra) [reply lib/exhal/navigation.ex]: I kinda like having this this function coordinate the work. Maybe we pull out the opt unpacking in to a private function? eg, `{tmpl_vars, opts} = interpret_follow_links_opts(opts)`

Comment (pezra) [reply lib/exhal/navigation.ex]: Hmm... do we even need the map on line 48? The previous implementation didn't unbox the `_follow_link` return values and the function docs indicate it returns a collection of `:ok`/`:error` tuples.

Comment (lmarlow) [reply lib/exhal/navigation.ex]: `Task.async_stream` wraps responses in `{:ok, value}` tuples, so we're just unwrapping that.

Comment (lmarlow) [reply lib/exhal/navigation.ex]: Pulled out into `tmpl_vars_and_opts_map/1`

File: lib/exhal/navigation.ex [hunk 2]
```diff
@@ -184,13 +180,41 @@ defmodule ExHal.Navigation do
...
+    end
+  end
+
   defp _follow_link(client, link, tmpl_vars, opts) do
     cond do
       Link.embedded?(link) ->
         {:ok, link.target, %ResponseHeader{status_code: 200}}
 
-      :else ->
+      true ->
```
Comment (pezra) [inline lib/exhal/navigation.ex]: :(

Comment (lmarlow) [reply lib/exhal/navigation.ex]: I was confused at how that worked... Oh... `:else` is truthy. I'll switch it back. I had never seen it before.

File: lib/exhal/navigation.ex [hunk 3]
```diff
@@ -12,9 +12,8 @@ defmodule ExHal.Navigation do
   `{:error, %ExHal.Error{...}}` if not
   """
   def follow_link(a_doc, name, opts \\ %{tmpl_vars: %{}, strict: false, headers: []}) do
-    opts = Map.new(opts)
+    {tmpl_vars, opts} = tmpl_vars_and_opts_map(opts)
     pick_volunteer? = !Map.get(opts, :strict, false)
```
Comment (pezra) [inline lib/exhal/navigation.ex]: We should put the `strict` extraction in `tmpl_vars_and_opts` too. Maybe rename it to `interpret_nav_opts` or something?

File: test/exhal_test.exs [hunk 4]
```diff
@@ -239,16 +271,16 @@ defmodule ExHalTest do
       end
     end
 
-    defp doc do
+    defp doc(scheme_and_domain \\ "http://example.com/") do
```
Comment (pezra) [inline test/exhal_test.exs]: Maybe `base_url` would better convey how this argument is used?

File: lib/exhal/navigation.ex [hunk 5]
```diff
@@ -184,6 +180,24 @@ defmodule ExHal.Navigation do
     end
   end
 
+  defp _follow_links(client, [link], tmpl_vars, opts) do
+    [_follow_link(client, link, tmpl_vars, opts)]
+  end
+
+  defp _follow_links(client, links, tmpl_vars, opts) do
+    {:ok, tsup} = Task.Supervisor.start_link()
```
Comment (pezra) [inline lib/exhal/navigation.ex]: Maybe this is a good argument for having exhal become an application?

File: test/exhal_test.exs (hunk 4 above)
Comment (lmarlow) [reply test/exhal_test.exs]: done

File: lib/exhal/navigation.ex (hunk 3 above)
Comment (lmarlow) [reply lib/exhal/navigation.ex]: done

File: lib/exhal/navigation.ex (hunk 5 above)
Comment (lmarlow) [reply lib/exhal/navigation.ex]: Let's see how it works before we push it to something with long running processes/supervisors. It's nice that this is just a functional library at the moment.

File: lib/exhal/navigation.ex [hunk 6]
```diff
@@ -184,13 +180,41 @@ defmodule ExHal.Navigation do
...
+
+    try do
+      Task.Supervisor.async_stream_nolink(tsup, links, &_follow_link(client, &1, tmpl_vars, opts))
```
[the remark this answers is not in the document]
Comment (lmarlow) [reply lib/exhal/navigation.ex]: done

File: lib/exhal/navigation.ex [hunk 7]
```diff
@@ -193,4 +197,36 @@ defmodule ExHal.Navigation do
         @client_module.get(client, Link.target_url!(link, tmpl_vars), opts)
     end
   end
+
+  @typep template_vars :: map()
+  @typep poison_options :: map()
+
+  @spec interpret_nav_opts(map()) :: {template_vars(), boolean(), poison_options()}
+  defp interpret_nav_opts(%{} = opts) do
+    {nav_options, poison_options} = Map.split(opts, [:tmpl_vars, :strict])
```
Comment (pezra) [inline lib/exhal/navigation.ex]: nice! i didn't know about `Map.split`.

File: lib/exhal/navigation.ex [hunk 8]
```diff
@@ -193,4 +197,36 @@ defmodule ExHal.Navigation do
         @client_module.get(client, Link.target_url!(link, tmpl_vars), opts)
     end
   end
+
+  @typep template_vars :: map()
+  @typep poison_options :: map()
```
Comment (pezra) [inline lib/exhal/navigation.ex]: i like the semantic types!

File: lib/exhal/navigation.ex (hunk 7 above)
Comment (lmarlow) [reply lib/exhal/navigation.ex]: Now we just need something like `Map.pop/3` that takes multiple keys somehow.

File: lib/exhal/navigation.ex (hunk 8 above)
Comment (lmarlow) [reply lib/exhal/navigation.ex]: 👍

**Comment by pezra** [comment]
released as v8.2.0

**Pull request merged**